Debug running dummy implementations

tylerdy opened the pull request:
## Debug running dummy implementations
<!-- Ideally, <=50 chars. 50 chars is here..: -->

### Description
<!-- For use in commit message, wrap at 72 chars. 72 chars is here: -->
- *Category*: <!-- one of bugfix, implementation, refactor, revert,
                   test, release, other/misc -->
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-XYZ
- *Research reference*: <!--Link to research documentation for code -->

### Changes and notes
<!-- 
Change description – why, what, anything unexplained by the above.
Include guidance to reviewers if changes are complex.
--> 
- Adds (WIP) pipeline_dummy.yaml and an input data yaml. 
- Fixes an output_slot in main pipeline schema constants

### Verification and Testing
<!--
Details on how code was verified. Consider: plots, images, (small) csv files.
-->
Ran using `easylink run -p pipeline_dummy.yaml -i src/easylink/steps/rl-dummy/input_data/input_data.yaml -e tests/specifications/common/environment_local.yaml`. May have to run `/ihme/homes/tylerdy/easylink/src/easylink/steps/rl-dummy/input_data/create_input_files.ipynb` to create input files beforehand.

aflaxman commented:
Cool!  I tried to run it on my laptop and ran into an issue with absolute paths that might be easy for you to fix:

```
(easylink-setup) abie@LT-MJ0L9VRK:~/easylink$ easylink run -p pipeline_dummy.yaml -i src/easylink/steps/rl-dummy/input_data/input_data.yaml -e tests/specifications/common/environment_local.yaml
2025-05-20 20:17:51.475 | 0:00:01.676140 | run:185 - Running pipeline
2025-05-20 20:17:51.476 | 0:00:01.676507 | run:187 - Results directory: /home/abie/easylink/results/2025_05_20_20_17_51
2025-05-20 20:17:51.668 | 0:00:01.868231 | exit_with_validation_error:141 -

==========================================
Validation errors found. Please see below.

INPUT DATA ERRORS:
  /ihme/homes/tylerdy/easylink/src/easylink/steps/rl-dummy/input_data/input_file_1.parquet:
  - File not found.
  /ihme/homes/tylerdy/easylink/src/easylink/steps/rl-dummy/input_data/input_file_2.parquet:
  - File not found.

Validation errors found. Please see above.
==========================================

(easylink-setup) abie@LT-MJ0L9VRK:~/easylink$ ls src/easylink/steps/rl-dummy/input_data/
create_input_files.ipynb  input_data.yaml  input_file_1.parquet  input_file_2.parquet
```

tylerdy commented:
> Cool! I tried to run it on my laptop and ran into an issue with absolute paths that might be easy for you to fix:

Thanks for catching this, I updated the paths to be on the simulation science drive. This is necessary rather than using a relative path to the input files in the repository bc of [this issue](https://jira.ihme.washington.edu/browse/MIC-6011).

I also removed some metadata entries that point to similarly inaccessible containers, so hopefully you should be able to run now (though it will still fail when executing the steps). Let me know how it goes!

zmbc commented:
@aflaxman This will not work on a laptop until we host the images somewhere accessible off-cluster.

zmbc commented:
@tylerdy Have you run this to the point where it generated a DAG SVG? It would be great to see what that looks like!

tylerdy commented:
> @tylerdy Have you run this to the point where it generated a DAG SVG? It would be great to see what that looks like!

![image](https://github.com/user-attachments/assets/f3d6e90b-ae55-4f71-bf1e-82b428f7e264)

patricktnast commented:
quoting: https://github.com/ihmeuw/easylink/pull/220#issuecomment-2899012990

> It all looks reasonable to me, except that I do not understand why there appear to be dependencies between the parallel splits for the validators?

interesting. That's probably a bug. @tylerdy can you please attach the snakefile to this thread?

tylerdy commented:
```
from easylink.utilities import validation_utils

rule all:
    message: 'Grabbing final output'
    localrule: True   
    input:
        final_output=['intermediate/dummy_canonicalizing_and_downstream_analysis/result.parquet'],
        validation='input_validations/final_validator',
    output: ['result.parquet']
    run:
        import os
        for input_path, output_path in zip(input.final_output, output):
            os.symlink(input_path, output_path)
rule:
    name: "results_main_input_validator"
    input: ['intermediate/dummy_canonicalizing_and_downstream_analysis/result.parquet']
    output: touch("input_validations/final_validator")
    localrule: True         
    message: "Validating results input slot main_input"
    run:
        for f in input:
            validation_utils.validate_input_file_dummy(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions_input_datasets_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/input_datasets_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions input slot input_datasets"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions_known_clusters_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/known_clusters_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions input slot known_clusters"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions"
    message: "Running determining_exclusions implementation: dummy_determining_exclusions" 
    input:
        input_datasets_and_input_known_clusters_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        known_clusters_and_maybe_input_datasets_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        validations=['input_validations/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/input_datasets_validator', 'input_validations/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/known_clusters_validator'],        
    output: ['intermediate/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/result.parquet']
    log: "diagnostics/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_determining_exclusions.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions
        export INPUT_DATASETS_AND_INPUT_KNOWN_CLUSTERS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export KNOWN_CLUSTERS_AND_MAYBE_INPUT_DATASETS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_1.parquet
        python /dummy_determining_exclusions.py > {log} 2>&1
        '''
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions_input_datasets_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/input_datasets_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions input slot input_datasets"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions_known_clusters_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/known_clusters_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions input slot known_clusters"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions"
    message: "Running determining_exclusions implementation: dummy_determining_exclusions" 
    input:
        input_datasets_and_input_known_clusters_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        known_clusters_and_maybe_input_datasets_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        validations=['input_validations/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/input_datasets_validator', 'input_validations/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/known_clusters_validator'],        
    output: ['intermediate/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/result.parquet']
    log: "diagnostics/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_determining_exclusions.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions
        export INPUT_DATASETS_AND_INPUT_KNOWN_CLUSTERS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export KNOWN_CLUSTERS_AND_MAYBE_INPUT_DATASETS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_2.parquet
        python /dummy_determining_exclusions.py > {log} 2>&1
        '''
rule:
    name: "dummy_clusters_to_links_known_clusters_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/dummy_clusters_to_links/known_clusters_validator")
    localrule: True         
    message: "Validating dummy_clusters_to_links input slot known_clusters"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "dummy_clusters_to_links"
    message: "Running clusters_to_links implementation: dummy_clusters_to_links" 
    input:
        known_clusters_and_maybe_input_datasets_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        validations=['input_validations/dummy_clusters_to_links/known_clusters_validator'],        
    output: ['intermediate/dummy_clusters_to_links/result.parquet']
    log: "diagnostics/dummy_clusters_to_links/dummy_clusters_to_links-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_clusters_to_links.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/dummy_clusters_to_links/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/dummy_clusters_to_links
        export KNOWN_CLUSTERS_AND_MAYBE_INPUT_DATASETS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        python /dummy_clusters_to_links.py > {log} 2>&1
        '''
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records_input_datasets_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/input_datasets_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records input slot input_datasets"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records_ids_to_remove_validator"
    input: ['intermediate/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/result.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/ids_to_remove_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records input slot ids_to_remove"
    run:
        for f in input:
            validation_utils.validate_ids_to_remove(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records"
    message: "Running removing_records implementation: dummy_removing_records" 
    input:
        input_datasets_and_input_known_clusters_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        ids_to_remove_file_path=['intermediate/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/result.parquet'],
        validations=['input_validations/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/input_datasets_validator', 'input_validations/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/ids_to_remove_validator'],        
    output: ['intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet']
    log: "diagnostics/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_removing_records.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records
        export INPUT_DATASETS_AND_INPUT_KNOWN_CLUSTERS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export IDS_TO_REMOVE_FILE_PATH=intermediate/determining_exclusions_and_removing_records_parallel_split_1_determining_exclusions_dummy_determining_exclusions/result.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_1.parquet
        python /dummy_removing_records.py > {log} 2>&1
        '''
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records_input_datasets_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/input_datasets_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records input slot input_datasets"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records_ids_to_remove_validator"
    input: ['intermediate/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/result.parquet']
    output: touch("input_validations/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/ids_to_remove_validator")
    localrule: True         
    message: "Validating determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records input slot ids_to_remove"
    run:
        for f in input:
            validation_utils.validate_ids_to_remove(f)
rule:
    name: "determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records"
    message: "Running removing_records implementation: dummy_removing_records" 
    input:
        input_datasets_and_input_known_clusters_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        ids_to_remove_file_path=['intermediate/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/result.parquet'],
        validations=['input_validations/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/input_datasets_validator', 'input_validations/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/ids_to_remove_validator'],        
    output: ['intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet']
    log: "diagnostics/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_removing_records.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records
        export INPUT_DATASETS_AND_INPUT_KNOWN_CLUSTERS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export IDS_TO_REMOVE_FILE_PATH=intermediate/determining_exclusions_and_removing_records_parallel_split_2_determining_exclusions_dummy_determining_exclusions/result.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_2.parquet
        python /dummy_removing_records.py > {log} 2>&1
        '''
rule:
    name: "pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing_datasets_validator"
    input: ['intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet', 'intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet']
    output: touch("input_validations/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/datasets_validator")
    localrule: True         
    message: "Validating pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing input slot datasets"
    run:
        for f in input:
            validation_utils.validate_dataset(f)
rule:
    name: "pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing"
    message: "Running pre-processing implementation: dummy_pre-processing" 
    input:
        dataset_file_paths=['intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet', 'intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet'],
        validations=['input_validations/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/datasets_validator'],        
    output: ['intermediate/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/result.parquet']
    log: "diagnostics/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_pre-processing.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing
        export DATASET_FILE_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet,intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_1.parquet
        python /dummy_pre-processing.py > {log} 2>&1
        '''
rule:
    name: "pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing_datasets_validator"
    input: ['intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet', 'intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet']
    output: touch("input_validations/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/datasets_validator")
    localrule: True         
    message: "Validating pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing input slot datasets"
    run:
        for f in input:
            validation_utils.validate_dataset(f)
rule:
    name: "pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing"
    message: "Running pre-processing implementation: dummy_pre-processing" 
    input:
        dataset_file_paths=['intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet', 'intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet'],
        validations=['input_validations/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/datasets_validator'],        
    output: ['intermediate/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/result.parquet']
    log: "diagnostics/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_pre-processing.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing
        export DATASET_FILE_PATHS=intermediate/determining_exclusions_and_removing_records_parallel_split_1_removing_records_dummy_removing_records/result.parquet,intermediate/determining_exclusions_and_removing_records_parallel_split_2_removing_records_dummy_removing_records/result.parquet
        export INPUT_DATASETS_SPLITTER_CHOICE=input_file_2.parquet
        python /dummy_pre-processing.py > {log} 2>&1
        '''
rule:
    name: "dummy_schema_alignment_datasets_validator"
    input: ['intermediate/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/result.parquet', 'intermediate/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/result.parquet']
    output: touch("input_validations/dummy_schema_alignment/datasets_validator")
    localrule: True         
    message: "Validating dummy_schema_alignment input slot datasets"
    run:
        for f in input:
            validation_utils.validate_dataset(f)
rule:
    name: "dummy_schema_alignment"
    message: "Running schema_alignment implementation: dummy_schema_alignment" 
    input:
        datasets_file_paths=['intermediate/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/result.parquet', 'intermediate/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/result.parquet'],
        validations=['input_validations/dummy_schema_alignment/datasets_validator'],        
    output: ['intermediate/dummy_schema_alignment/result.parquet']
    log: "diagnostics/dummy_schema_alignment/dummy_schema_alignment-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_schema_alignment.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/dummy_schema_alignment/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/dummy_schema_alignment
        export DATASETS_FILE_PATHS=intermediate/pre-processing_parallel_split_1_pre-processing_parallel_split_1_dummy_pre-processing/result.parquet,intermediate/pre-processing_parallel_split_2_pre-processing_parallel_split_2_dummy_pre-processing/result.parquet
        python /dummy_schema_alignment.py > {log} 2>&1
        '''
rule:
    name: "splink_blocking_and_filtering_known_links_validator"
    input: ['intermediate/dummy_clusters_to_links/result.parquet']
    output: touch("input_validations/splink_blocking_and_filtering/known_links_validator")
    localrule: True         
    message: "Validating splink_blocking_and_filtering input slot known_links"
    run:
        for f in input:
            validation_utils.validate_links(f)
rule:
    name: "splink_blocking_and_filtering_records_validator"
    input: ['intermediate/dummy_schema_alignment/result.parquet']
    output: touch("input_validations/splink_blocking_and_filtering/records_validator")
    localrule: True         
    message: "Validating splink_blocking_and_filtering input slot records"
    run:
        for f in input:
            validation_utils.validate_records(f)
rule:
    name: "splink_blocking_and_filtering"
    message: "Running blocking_and_filtering implementation: splink_blocking_and_filtering" 
    input:
        known_links_file_path=['intermediate/dummy_clusters_to_links/result.parquet'],
        records_file_path=['intermediate/dummy_schema_alignment/result.parquet'],
        validations=['input_validations/splink_blocking_and_filtering/known_links_validator', 'input_validations/splink_blocking_and_filtering/records_validator'],        
    output: directory('intermediate/splink_blocking_and_filtering/blocks')
    log: "diagnostics/splink_blocking_and_filtering/splink_blocking_and_filtering-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/zmbc/splink_blocking_and_filtering.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/splink_blocking_and_filtering/blocks
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/splink_blocking_and_filtering
        export KNOWN_LINKS_FILE_PATH=intermediate/dummy_clusters_to_links/result.parquet
        export RECORDS_FILE_PATH=intermediate/dummy_schema_alignment/result.parquet
        python /splink_blocking_and_filtering.py > {log} 2>&1
        '''
rule:
    name: "splink_evaluating_pairs_known_links_validator"
    input: ['intermediate/dummy_clusters_to_links/result.parquet']
    output: touch("input_validations/splink_evaluating_pairs/known_links_validator")
    localrule: True         
    message: "Validating splink_evaluating_pairs input slot known_links"
    run:
        for f in input:
            validation_utils.validate_links(f)
rule:
    name: "splink_evaluating_pairs_blocks_validator"
    input: ['intermediate/splink_blocking_and_filtering/blocks']
    output: touch("input_validations/splink_evaluating_pairs/blocks_validator")
    localrule: True         
    message: "Validating splink_evaluating_pairs input slot blocks"
    run:
        for f in input:
            validation_utils.validate_blocks(f)
rule:
    name: "splink_evaluating_pairs"
    message: "Running evaluating_pairs implementation: splink_evaluating_pairs" 
    input:
        known_links_file_path=['intermediate/dummy_clusters_to_links/result.parquet'],
        blocks_dir_path=['intermediate/splink_blocking_and_filtering/blocks'],
        validations=['input_validations/splink_evaluating_pairs/known_links_validator', 'input_validations/splink_evaluating_pairs/blocks_validator'],        
    output: ['intermediate/splink_evaluating_pairs/result.parquet']
    log: "diagnostics/splink_evaluating_pairs/splink_evaluating_pairs-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/zmbc/splink_evaluating_pairs.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/splink_evaluating_pairs/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/splink_evaluating_pairs
        export KNOWN_LINKS_FILE_PATH=intermediate/dummy_clusters_to_links/result.parquet
        export BLOCKS_DIR_PATH=intermediate/splink_blocking_and_filtering/blocks
        python /splink_evaluating_pairs.py > {log} 2>&1
        '''
rule:
    name: "splink_links_to_clusters_links_validator"
    input: ['intermediate/splink_evaluating_pairs/result.parquet']
    output: touch("input_validations/splink_links_to_clusters/links_validator")
    localrule: True         
    message: "Validating splink_links_to_clusters input slot links"
    run:
        for f in input:
            validation_utils.validate_links(f)
rule:
    name: "splink_links_to_clusters"
    message: "Running links_to_clusters implementation: splink_links_to_clusters" 
    input:
        links_file_path=['intermediate/splink_evaluating_pairs/result.parquet'],
        validations=['input_validations/splink_links_to_clusters/links_validator'],        
    output: ['intermediate/splink_links_to_clusters/result.parquet']
    log: "diagnostics/splink_links_to_clusters/splink_links_to_clusters-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/zmbc/splink_links_to_clusters.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/splink_links_to_clusters/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/splink_links_to_clusters
        export LINKS_FILE_PATH=intermediate/splink_evaluating_pairs/result.parquet
        python /splink_links_to_clusters.py > {log} 2>&1
        '''
rule:
    name: "dummy_updating_clusters_known_clusters_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/dummy_updating_clusters/known_clusters_validator")
    localrule: True         
    message: "Validating dummy_updating_clusters input slot known_clusters"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "dummy_updating_clusters_new_clusters_validator"
    input: ['intermediate/splink_links_to_clusters/result.parquet']
    output: touch("input_validations/dummy_updating_clusters/new_clusters_validator")
    localrule: True         
    message: "Validating dummy_updating_clusters input slot new_clusters"
    run:
        for f in input:
            validation_utils.validate_clusters(f)
rule:
    name: "dummy_updating_clusters"
    message: "Running updating_clusters implementation: dummy_updating_clusters" 
    input:
        known_clusters_and_maybe_input_datasets_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        new_clusters_file_path=['intermediate/splink_links_to_clusters/result.parquet'],
        validations=['input_validations/dummy_updating_clusters/known_clusters_validator', 'input_validations/dummy_updating_clusters/new_clusters_validator'],        
    output: ['intermediate/dummy_updating_clusters/result.parquet']
    log: "diagnostics/dummy_updating_clusters/dummy_updating_clusters-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_updating_clusters.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/dummy_updating_clusters/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/dummy_updating_clusters
        export KNOWN_CLUSTERS_AND_MAYBE_INPUT_DATASETS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export NEW_CLUSTERS_FILE_PATH=intermediate/splink_links_to_clusters/result.parquet
        python /dummy_updating_clusters.py > {log} 2>&1
        '''
rule:
    name: "dummy_canonicalizing_and_downstream_analysis_input_datasets_validator"
    input: ['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet']
    output: touch("input_validations/dummy_canonicalizing_and_downstream_analysis/input_datasets_validator")
    localrule: True         
    message: "Validating dummy_canonicalizing_and_downstream_analysis input slot input_datasets"
    run:
        for f in input:
            validation_utils.validate_input_dataset_or_known_clusters(f)
rule:
    name: "dummy_canonicalizing_and_downstream_analysis_clusters_validator"
    input: ['intermediate/dummy_updating_clusters/result.parquet']
    output: touch("input_validations/dummy_canonicalizing_and_downstream_analysis/clusters_validator")
    localrule: True         
    message: "Validating dummy_canonicalizing_and_downstream_analysis input slot clusters"
    run:
        for f in input:
            validation_utils.validate_clusters(f)
rule:
    name: "dummy_canonicalizing_and_downstream_analysis"
    message: "Running canonicalizing_and_downstream_analysis implementation: dummy_canonicalizing_and_downstream_analysis" 
    input:
        input_datasets_and_input_known_clusters_file_paths=['/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet', '/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet'],
        clusters_file_path=['intermediate/dummy_updating_clusters/result.parquet'],
        validations=['input_validations/dummy_canonicalizing_and_downstream_analysis/input_datasets_validator', 'input_validations/dummy_canonicalizing_and_downstream_analysis/clusters_validator'],        
    output: ['intermediate/dummy_canonicalizing_and_downstream_analysis/result.parquet']
    log: "diagnostics/dummy_canonicalizing_and_downstream_analysis/dummy_canonicalizing_and_downstream_analysis-output.log"
    container: "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/tylerdy/dummy_canonicalizing_and_downstream_analysis.sif" 
    shell:
        '''
        export DUMMY_CONTAINER_OUTPUT_PATHS=intermediate/dummy_canonicalizing_and_downstream_analysis/result.parquet
        export DUMMY_CONTAINER_DIAGNOSTICS_DIRECTORY=diagnostics/dummy_canonicalizing_and_downstream_analysis
        export INPUT_DATASETS_AND_INPUT_KNOWN_CLUSTERS_FILE_PATHS=/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet,/mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
        export CLUSTERS_FILE_PATH=intermediate/dummy_updating_clusters/result.parquet
        python /dummy_canonicalizing_and_downstream_analysis.py > {log} 2>&1
        '''
```

zmbc commented:
**TL;DR: The diagram above is correct and we don't think there is a bug.**

Notes from a live conversation with @patricktnast about this:
- The dependencies I was concerned about are between two **distinct** cloneable sections/ParallelSteps: the one in [this diagram](https://easylink.readthedocs.io/en/latest/concepts/pipeline_schema/index.html#entity-resolution-sub-steps) and the one in [this diagram](https://easylink.readthedocs.io/en/latest/concepts/pipeline_schema/index.html#linking-sub-steps). When you unravel the hierarchy to get to the implementation level, these sections are actually adjacent to one another.
- The names in the diagram (though incredibly confusing) actually indicate this: there is the `determining_exclusions_and_removing_records_parallel_split` and the `pre-processing_parallel_split`.
- Because of our workaround, of passing all data to each clone of a cloneable section/ParallelStep and leaving it up to the implementation to select the relevant part, it is correct that the outputs of both of the `determining_exclusions_and_removing_records_parallel_split` splits/clones would go to both of the `determining_exclusions_and_removing_records_parallel_split` splits/clones.

tylerdy commented:
>     it is correct that the outputs of both of the `determining_exclusions_and_removing_records_parallel_split` splits/clones would go to both of the `determining_exclusions_and_removing_records_parallel_split` splits/clones.

Do you mean to the pre-processing parallel splits/clones?

zmbc commented:
Oops, yes -- corrected.

tylerdy commented:
Currently able to run all the way to the final ouput validator which is still trying to validate based on the development schema input file structure - may be EasyLink bug. 

Aside from resolving that, also still need to add a `validate_dataset_dir_list` validator.

tylerdy commented:
After merging Zeb's fix in @225 I'm now able to run the entire EasyLink pipeline!

zmbc commented on src/easylink/steps/rl-dummy/input_data/input_data.yaml:
```diff
@@ -0,0 +1,3 @@
+input_file_1: /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_1.parquet
+input_file_2: /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/input_file_2.parquet
+input_file_3: /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
```
```suggestion
known_clusters: /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/rl-dummy/known_clusters.parquet
```

I realize it doesn't make any difference, but this seems a lot less confusing

zmbc commented on src/easylink/steps/rl-dummy/pre-processing/dummy_pre-processing.py:
```diff
@@ -25,22 +27,30 @@ def load_file(file_path, file_format=None):
 
 # DATASET_FILE_PATHS is list of filepaths
 dataset_paths = os.environ["DATASET_FILE_PATHS"].split(",")
+logging.info(f"dataset_paths={dataset_paths}")
```
This is totally fine as-is but here's a cool f-string trick you might not know about: this is equivalent to what you have
```suggestion
logging.info(f"{dataset_paths=}")
```

zmbc commented on src/easylink/steps/rl-dummy/schema_alignment/dummy_schema_alignment.py:
```diff
@@ -24,11 +25,17 @@ def load_file(file_path, file_format=None):
+for dir in dataset_dirs:
+    for root, dirs, files in os.walk(dir):
+        for file in files:
+            if file == ".snakemake_timestamp":
```
```suggestion
            if file.startswith("."):
```

zmbc commented on src/easylink/steps/splink/splink_evaluating_pairs.py:
```diff
@@ -12,10 +12,13 @@
 all_predictions = []
 
 for block_dir in blocks_dir.iterdir():
+    if "snakemake" in str(block_dir):
```
Is this also checking for the timestamp? Let's change this to be `.startswith(".")` as well if so

zmbc commented on src/easylink/utilities/validation_utils.py:
```diff
@@ -373,6 +373,23 @@ def validate_dir(filepath: str) -> None:
+    if not input_path.is_dir():
+        raise NotADirectoryError(f"The path {filepath} is not a directory.")
+
+    file_path = ""
+    for f in input_path.iterdir():
+        if "snakemake" in str(f):
+            continue
+        if file_path != "":
+            raise ValueError(f"The directory {input_path} contains more than one file.")
+        file_path = f
+    if file_path == "":
+        raise FileNotFoundError(f"The directory {input_path} does not contain any files.")
```
```suggestion
    file_paths = [f for f in input_path.iterdir() if not f.startswith(".")]
    if len(file_paths) > 1:
      raise ValueError(f"The directory {input_path} contains more than one file.")
    if len(file_paths) == 0:
      raise FileNotFoundError(f"The directory {input_path} does not contain any files.")
    
    file_path = file_paths[0]
```

just stylistic changes

tylerdy commented on pipeline_dummy.yaml:
```diff
@@ -0,0 +1,69 @@
+                  name: dummy_schema_alignment
+              blocking_and_filtering:
+                implementation:
+                  name: splink_blocking_and_filtering
```
Didn't think about this when we were talking earlier but the only implementation with both a dummy and splink implementation is links_to_clusters, which I don't think I ever tested bc I couldn't think of a reason to ever use it vs the splink implementation (the dummy more or less just assigns random records to random clusters). For the same reasons I don't think it makes sense to write dummy implementations for the other steps with splink implementations.

So, I won't plan to add another e2e test that uses less/no splink implementations, unless you know of reasons I should.

zmbc replied:
That makes sense! I think in that case you could just remove that implementation (`dummy_links_to_clusters`)

zmbc commented on tests/e2e/test_pipeline_splink_dummy.py:
```diff
@@ -0,0 +1,92 @@
+
+from easylink.utilities.general_utils import is_on_slurm
+
+RESULT_CHECKSUM = "51496c06439823dd483bb43a016dcf07c014a4ccc5b09a9cc98c1b99f404b19f"
+
```
```suggestion
```

zmbc commented on tests/e2e/test_pipeline_splink_dummy.py:
```diff
@@ -0,0 +1,92 @@
+        assert final_output.exists()
+        # Check that the results file checksum matches the expected value
+        results = pd.read_parquet(final_output).sort_values("Input Record ID").reset_index()
+        results["index"] = results.index
```
It might be a bit nicer to remove this "column" from the CSV file (when you save a Pandas dataframe as a CSV, it will save the index as another column, which is rarely what you want). But not a big deal either way, you could just add a note.

zmbc commented on tests/e2e/test_pipeline_splink_dummy.py:
```diff
@@ -0,0 +1,92 @@
+        )
+        final_output = test_specific_results_dir / "result.parquet"
+        assert final_output.exists()
+        # Check that the results file checksum matches the expected value
```
```suggestion
        # Check that the results file matches the expected value
```

zmbc reviewed (approved):
Great work! (I am assuming the test passes!)
